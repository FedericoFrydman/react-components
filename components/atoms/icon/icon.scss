@import '../../assets/styles/brand.scss';
@import '../../assets/styles/colors.scss';

.lsux-icon-container {
  display: flex;
  flex-wrap: wrap;
}

.lsux-icon {
  filter: $filter-default;
  transition: all 0.15s ease-in-out;

  &--no-transition {
    transition: none;
  }

  &.lsux-icon--mono {
    filter: none;
  }

  &--clickable {
    cursor: pointer;
  }

  &--xsmall {
    height: 12px;
    width: 12px;
  }

  &--small {
    height: 16px;
    width: 16px;
  }

  &--medium-small {
    height: 18px;
    width: 18px;
  }

  &--medium {
    height: 24px;
    width: 24px;
  }

  &--large {
    height: 32px;
    width: 32px;
  }

  &--xlarge {
    height: 48px;
    width: 48px;
  }

  &--N00 {
    filter: $filter-white;
  }
  &--N100 {
    filter: $filter-neutral-100;
  }
  &--N300 {
    filter: $filter-neutral-300;
  }
  &--N500 {
    filter: $filter-light-gray;
  }
  &--N700 {
    filter: $filter-gray;
  }
  &--N800 {
    filter: $filter-black;
  }
  &--N1000 {
    filter: $filter-deep-black;
  }

  &--G10 {
    filter: $filter-green-10;
  }
  &--G100 {
    filter: $filter-green-100;
  }
  &--G200 {
    filter: $filter-green;
  }
  &--G300 {
    filter: $filter-green-300;
  }

  &--P10 {
    filter: $filter-purple-10;
  }
  &--P100 {
    filter: $filter-purple-100;
  }
  &--P200 {
    filter: $filter-purple-200;
  }
  &--P300 {
    filter: $filter-purple-300;
  }

  &--R10 {
    filter: $filter-red-10;
  }
  &--R100 {
    filter: $filter-red-100;
  }
  &--R200 {
    filter: $filter-red-200;
  }
  &--R300 {
    filter: $filter-red;
  }

  &--B10 {
    filter: $filter-blue-10;
  }
  &--B100 {
    filter: $filter-blue-100;
  }
  &--B200 {
    filter: $filter-blue;
  }
  &--B300 {
    filter: $filter-blue-300;
  }

  &--Y10 {
    filter: $filter-yellow-10;
  }
  &--Y100 {
    filter: $filter-yellow-100;
  }
  &--Y200 {
    filter: $filter-yellow-200;
  }
  &--Y300 {
    filter: $filter-yellow;
  }

  &--BRAND50 {
    filter: var(--brand-color-prim50-filter);
  }

  &--BRAND100 {
    filter: var(--brand-color-prim100-filter);
  }

  &--BRAND200 {
    filter: var(--brand-color-prim200-filter);
  }

  &--BRAND300 {
    filter: var(--brand-color-prim300-filter);
  }
}

.circle {
  background: $R200;
  border-radius: 50%;

  &--xsmall {
    height: 4px;
    width: 4px;
  }

  &--small {
    height: 5px;
    width: 5px;
  }

  &--medium-small {
    height: 6px;
    width: 6px;
  }

  &--medium {
    height: 7px;
    width: 7px;
  }

  &--large {
    height: 9px;
    width: 9px;
  }

  &--xlarge {
    height: 11px;
    width: 11px;
  }
}
