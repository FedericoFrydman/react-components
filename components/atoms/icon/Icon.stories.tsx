import { Story, Meta } from '@storybook/react';

import { Icon, IconProps } from './Icon';

import docs from './Icon.docs.mdx';
import { ICON_NAMES } from '../../../components/assets/icons';
import { ICON_COLORS_V2 } from '../../../components/assets/icons/icons-v2';
import { Container } from '../container/Container';

export const Default: Story<IconProps> = (args: IconProps) => (
  <Container classNames={['m-3']}>
    <Icon name="action_add" {...args} />
  </Container>
);

export const Clickable = Default.bind({});
Clickable.args = {
  name: 'action_add',
  onClick: () => alert('Add icon clicked'),
};

export const ExtraLargeIconWithBadge = Default.bind({});
ExtraLargeIconWithBadge.args = {
  name: 'action_add',
  size: 'xlarge',
  useBadge: true,
};

export const LargeIconWithBadge = Default.bind({});
LargeIconWithBadge.args = {
  name: 'action_add',
  size: 'large',
  useBadge: true,
};

export const MediumIconWithBadge = Default.bind({});
MediumIconWithBadge.args = {
  name: 'action_add',
  size: 'medium',
  useBadge: true,
};

export const SmallIconWithBadge = Default.bind({});
SmallIconWithBadge.args = {
  name: 'action_add',
  size: 'small',
  useBadge: true,
};

export const XsmallIconWithBadge = Default.bind({});
XsmallIconWithBadge.args = {
  name: 'action_add',
  size: 'xsmall',
  useBadge: true,
};

export default {
  argTypes: {
    color: {
      control: {
        type: 'select',
      },
      options: ICON_COLORS_V2,
    },
    name: { control: { type: 'select' }, options: ICON_NAMES },
    noTransition: { control: { type: 'boolean' } },
    size: { control: { type: 'select' }, options: ['xsmall', 'small', 'medium-small', 'medium', 'large', 'xlarge'] },
    useBadge: { control: { type: 'boolean' } },
  },
  component: Default,
  parameters: {
    docs: { page: docs },
  },
  title: 'Components/Icon',
} as Meta;
