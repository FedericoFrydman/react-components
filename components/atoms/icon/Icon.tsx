import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './icon.scss';
import { Container } from '../container/Container';
import { Icons } from '../../assets/icons';
import { IconColorsV1, IconNamesV1 } from '../../assets/icons/icons-v1';
import { IconColorsV2, IconNamesV2 } from '../../assets/icons/icons-v2';

export type IconLibraryVersion = '1' | '2';

export type IconNames = IconNamesV1 | IconNamesV2;

export type IconAndBadgeSizes = 'xsmall' | 'small' | 'medium-small' | 'medium' | 'large' | 'xlarge';

export const excludedCategories: string[] = ['apps', 'badges', 'logo', 'logos', 'feedback', 'payments', 'medialogo'];

export interface IconProps
  extends Omit<React.HTMLProps<HTMLImageElement>, 'css' | 'alt' | 'src' | 'size' | 'crossOrigin'> {
  /**
   * Applies a color filter to the icon.
   */
  color?: IconColorsV1 | IconColorsV2;
  /**
   * Name of the icon to be shown.
   */
  name: IconNamesV1 | IconNamesV2;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Defines icon size and badge size.
   */
  size?: IconAndBadgeSizes;
  /**
   * Disables CSS transition on the icon.
   */
  noTransition?: boolean;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  onClick?: (e: React.MouseEvent) => void | unknown;
  /**
   * Alt tag for the HTML element
   * @default `name` paramater
   */
  alt?: string;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLImageElement>;
  /**
   * Define if icon has a Badge.
   */
  useBadge?: boolean;
}

const IconComponent = (
  {
    classNames = [],
    color = 'default',
    name,
    alt = name + '.',
    onClick,
    size = 'medium',
    noTransition,
    useBadge = false,
    ...props
  }: IconProps,
  ref: IconProps['ref'],
) => {
  const classIcon = 'lsux-icon';
  const classIconContainer = 'lsux-icon-container';

  const applyMonoClass = () => {
    let isMono = false;
    excludedCategories.forEach((c) => {
      if (name?.includes(`${c}_`)) isMono = true;
    });
    return isMono ? `${classIcon}--mono` : '';
  };

  const cnIconContainer = [`${classIconContainer}`, `flex-row`];

  const cnIcon = cn(
    {
      [`${classIcon}`]: true,
      [`${classIcon}--${color}`]: color !== 'default',
      [`${classIcon}--${size}`]: size,
      [`${classIcon}--no-transition`]: noTransition,
      [`${classIcon}--clickable`]: !!onClick,
      [applyMonoClass()]: true,
    },
    classNames,
  );

  const cnBadge = cn({
    [`circle`]: true,
    [`circle--${size}`]: size,
  });

  if (name === null || name === undefined) {
    return <div onClick={onClick} className={cnIcon}></div>;
  } else {
    return (
      <Container classNames={cnIconContainer} background="none" flexbox flexDirection="row">
        <img {...props} src={Icons[name] as string} alt={alt} className={cnIcon} ref={ref} onClick={onClick}></img>
        {useBadge && <div className={cnBadge} />}
      </Container>
    );
  }
};

export const Icon = forwardRef<HTMLImageElement, IconProps>(IconComponent);
