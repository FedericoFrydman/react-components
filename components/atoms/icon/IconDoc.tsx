import { FC } from 'react';
import { ICON_NAMES_V1 } from '../../assets/icons/icons-v1';
import { ICON_NAMES_V2 } from '../../assets/icons/icons-v2';
import { Heading } from '../heading/Heading';
import { Icon, IconLibraryVersion } from './Icon';
import { Text } from '../text/Text';

const groupByCategories = (libraryVersion: IconLibraryVersion): Record<string, string[]>[] => {
  const category: Record<string, string[]> = {};
  const result: Record<string, string[]>[] = [];

  const ICON_NAMES = libraryVersion === '1' ? ICON_NAMES_V1 : ICON_NAMES_V2;

  ICON_NAMES.forEach((icon) => {
    let key: string = icon.split('_')[0];

    if (libraryVersion === '1') {
      switch (key) {
        case 'action':
          key = 'Action';
          break;
        case 'alert':
          key = 'Alert';
          break;
        case 'comm':
          key = 'Communication';
          break;
        case 'cs':
          key = 'Commercial Shield';
          break;
        case 'file':
          key = 'File';
          break;
        case 'id':
          key = 'IDShield';
          break;
        case 'logo':
          key = 'Logos';
          break;
        case 'ls':
          key = 'Legalshield';
          break;
        case 'nav':
          key = 'Navigation';
          break;
        case 'object':
          key = 'Object';
          break;
        case 'payment':
          key = 'Payment';
          break;
        default:
          break;
      }
    } else if (libraryVersion === '2') {
      switch (key) {
        case 'apps':
          key = 'Apps';
          break;
        case 'arrows':
          key = 'Arrows';
          break;
        case 'badges':
          key = 'Badges';
          break;
        case 'calendar':
          key = 'Calendar';
          break;
        case 'comm':
          key = 'Communication';
          break;
        case 'edit':
          key = 'Edit';
          break;
        case 'feedback':
          key = 'Feedback';
          break;
        case 'file':
          key = 'File';
          break;
        case 'filetypes':
          key = 'File Types';
          break;
        case 'interface':
          key = 'Interface';
          break;
        case 'live':
          key = 'Live';
          break;
        case 'logos':
          key = 'Logos';
          break;
        case 'marks':
          key = 'Marks';
          break;
        case 'media':
          key = 'Media';
          break;
        case 'medialogo':
          key = 'Media Logos';
          break;
        case 'menu':
          key = 'Menu';
          break;
        case 'nav':
          key = 'Navigation';
          break;
        case 'partnerships':
          key = 'Partnerships';
          break;
        case 'payments':
          key = 'Payments';
          break;
        case 'social':
          key = 'Social';
          break;
        case 'system':
          key = 'System';
          break;
        case 'user':
          key = 'User';
          break;
        case 'warning':
          key = 'Warning';
          break;
        default:
          break;
      }
    }

    if (!category[key]) {
      category[key] = [];
      result.push({ [key]: category[key] });
    }
    category[key].push(icon);
  });
  return result;
};

export const IconDoc: FC<{ libraryVersion: IconLibraryVersion }> = ({
  libraryVersion,
}: {
  libraryVersion: IconLibraryVersion;
}) => {
  return (
    <>
      {groupByCategories(libraryVersion).map((category) => {
        const title = Object.keys(category)[0];
        return (
          <div key={title} className="mb-3">
            <Heading as="T16" text={title} />
            <div style={{ display: 'flex', flexWrap: 'wrap' }}>
              {category[title].map((icon) => {
                return (
                  <div key={icon} className="m-5" style={{ textAlign: 'center', width: '7.5rem' }}>
                    <Icon id={icon} size="large" name={icon} />
                    <label htmlFor={icon} style={{ display: 'block' }}>
                      <Text textSize="small" text={icon} />
                    </label>
                  </div>
                );
              })}
            </div>
          </div>
        );
      })}
    </>
  );
};
