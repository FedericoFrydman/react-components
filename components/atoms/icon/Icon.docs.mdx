import { Meta, ArgsTable, Story, Canvas, Title, Description } from '@storybook/addon-docs';

import { IconDoc } from './IconDoc';
import { Icon } from './Icon';
import { Heading } from '../heading/Heading';

<Title />

<Description />

<Canvas>
  <Story id="components-icon--default" />
</Canvas>

<Meta />

<ArgsTable of={Icon} />

## Sizes

| <Heading as="T16" text="XSmall" />          | <Heading as="T16" text="Small" />          | <Heading as="T16" text="Medium Small" />          | <Heading as="T16" text="Medium" />          | <Heading as="T16" text="Large" />          | <Heading as="T16" text="XLarge" />          |
| ------------------------------------------- | ------------------------------------------ | ------------------------------------------------- | ------------------------------------------- | ------------------------------------------ | ------------------------------------------- |
| <Heading as="T16" text="12x12 px" />        | <Heading as="T16" text="16x16 px" />       | <Heading as="T16" text="18x18 px" />              | <Heading as="T16" text="24x24 px" />        | <Heading as="T16" text="32x32 px" />       | <Heading as="T16" text="48x48 px" />        |
| <Icon name="system_camera" size="xsmall" /> | <Icon name="system_camera" size="small" /> | <Icon name="system_camera" size="medium-small" /> | <Icon name="system_camera" size="medium" /> | <Icon name="system_camera" size="large" /> | <Icon name="system_camera" size="xlarge" /> |

## Colors

After version v1.11.0 icons have the option to use any color in our [color palette](?path=/docs/foundation-colors--page)

**For old color references:**

| Old color name | New color name |
| -------------- | -------------- |
| blue           | B200           |
| green          | G200           |
| light-gray     | N500           |
| gray           | N700           |
| black          | N800           |
| deep-black     | N1000          |
| red            | R300           |
| yellow         | Y300           |

These is the list for all colors now available for ICON_COLORS

| Color name | Palette color reference | HEX code |
| ---------- | ----------------------- | -------- |
| default    |
| B10        | $B10                    | #e9eff9  |
| B100       | $B100                   | #377bdd  |
| B200       | $B200                   | #1952bd  |
| B200       | $B200                   | #13429a  |
| G10        | $G10                    | #e9f4ed  |
| G100       | $G100                   | #34ab5e  |
| G200       | $G200                   | #2c9150  |
| G300       | $G300                   | #006e3c  |
| N00        | $N00                    | #ffffff  |
| N100       | $N100                   | #f7f7f7  |
| N300       | $N300                   | #eeeeee  |
| N500       | $N500                   | #afafaf  |
| N700       | $N700                   | #545454  |
| N800       | $N800                   | #222222  |
| N1000      | $N1000                  | #000000  |
| P10        | $P10                    | #f0ecf3  |
| P100       | $P100                   | #a073bc  |
| P200       | $P200                   | #70468c  |
| P300       | $P300                   | #431c5e  |
| R10        | $R10                    | #ffeef0  |
| R100       | $R100                   | #ce0220  |
| R200       | $R200                   | #b5021c  |
| R300       | $R300                   | #920217  |
| Y10        | $Y10                    | #fffee6  |
| Y100       | $Y100                   | #ffd633  |
| Y200       | $Y200                   | #ffcc00  |
| Y300       | $Y300                   | #e6b800  |
| BRAND50    | name                    | NA       |
| BRAND100   | name                    | NA       |
| BRAND200   | name                    | NA       |
| BRAND300   | name                    | NA       |

## Events

An icon has all the properties of an img tag, so it can receive events. Particularly the **onClick** event adds the `cursor: pointer` style.

<Canvas>
  <Story id="components-icon--clickable" />
</Canvas>

## Icons

Below are icons use included for responsive web available in different icon colors and icon sizes. For iOS and Android app, "System" icons are the set from the OS itself. We took advantage to use them for a cohesive app experience and easy to maintenance.

## Version 2 Icons

<IconDoc libraryVersion="2" />

## Version 1 Icons

<IconDoc libraryVersion="1" />
