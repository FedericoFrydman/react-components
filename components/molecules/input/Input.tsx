import { cloneElement, DetailedHTMLProps, forwardRef, ReactElement, Ref, useEffect, useState } from 'react';
import NumberFormat, { NumberFormatProps } from 'react-number-format';
import cn from 'classnames';

import { Button } from '../button/Button';
import { Icon, IconNames, IconProps } from '../../atoms/icon/Icon';
import { IconColors } from '../../../components/assets/icons';
import './input.scss';

export type InputValidationStatus = '' | 'valid' | 'warning' | 'invalid';

export type InputType =
  | 'color'
  | 'date'
  | 'datetime-local'
  | 'email'
  | 'hidden'
  | 'month'
  | 'number'
  | 'password'
  | 'range'
  | 'search'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week'
  | 'currency';

export interface InputProps extends Omit<React.HTMLProps<HTMLInputElement>, 'type'> {
  /**
   * Triggers error | warning | success styles on the component. Important for accessibility.
   */
  status?: InputValidationStatus;
  /**
   * Determines the type of the Input.
   * @default text
   */
  type?: InputType;
  /**
   * Determines which icon should be shown on the right; if a password type this property is managed internally with the hide/show icons.
   * @deprecated Please use `iconLeft` and `iconRight` instead.
   */
  icon?: IconNames;
  /**
   * Icon on the left side of the input
   */
  iconLeft?: ReactElement<IconProps>;
  /**
   * Icon on the right side of the input
   */
  iconRight?: ReactElement<IconProps>;
  /**
   * Applies a color to the icon.  This could be for the icon on the right or a search icon when type is search.
   * @default BRAND200
   */
  iconColor?: IconColors;
  /**
   * Custom class names for the Input.
   */
  classNames?: cn.Argument;
  /**
   * Format uses the NumberFormat component instead of Input. It allows for input masking.
   */
  format?: string;
  /**
   * Placeholder value for the input field.
   */
  placeholder?: string;
  /**
   * Prefix value for the input field. Only displays when `NumberFormat` component is rendered.
   */
  prefix?: string;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLInputElement>;
}

const InputComponent = (
  {
    status = '',
    type = 'text',
    icon,
    iconColor = 'BRAND200',
    format,
    placeholder,
    prefix,
    iconLeft,
    iconRight = icon ? <Icon name={icon} color={iconColor} /> : null,
    classNames = [],
    ...props
  }: InputProps,
  ref: InputProps['ref'],
) => {
  const [iconColorState, setIconColorState] = useState(iconColor);
  const [passwordVisible, setPasswordVisible] = useState(false);

  const classInputContainer = `lsux-input-container`;
  const cnInputContainer = cn(classInputContainer);

  const classInput = 'lsux-input-container__input';
  const cnInput = cn(
    {
      [`${classInput}`]: true,
      [`${classInput}--${status}`]: status,
      [`${classInput}--icon-left`]: type === 'search' || iconLeft,
      [`${classInput}--icon-right`]: type === 'password' || iconRight,
    },
    classNames,
  );

  const classInputIcon = 'lsux-input__icon';

  useEffect(() => {
    switch (status) {
      case 'invalid': {
        setIconColorState('R200');
        break;
      }
      case 'warning': {
        setIconColorState('Y200');
        break;
      }
      case 'valid': {
        setIconColorState('G200');
        break;
      }
      default: {
        setIconColorState(iconColor);
        break;
      }
    }
  }, [status, iconColor]);

  if (type === 'password' && !iconColor) {
    setIconColorState('BRAND200');
  }

  if (iconLeft) {
    iconLeft = cloneElement(iconLeft, {
      classNames: [iconLeft.props?.classNames, classInputIcon, `${classInputIcon}-left`],
      color: iconColorState,
    });
  }

  if (iconRight) {
    iconRight = cloneElement(iconRight, {
      classNames: [iconRight.props?.classNames, classInputIcon, `${classInputIcon}-right`],
      color: iconColorState,
    });
  }

  if (type === 'search') {
    iconLeft = (
      <Icon
        name="interface_search_magnifying_glass"
        classNames={[classInputIcon, `${classInputIcon}-search-icon`]}
        color={iconColorState}
      />
    );
  }

  if (type != 'password' && icon) {
    iconRight = <Icon name={icon} color={iconColorState} classNames={classInputIcon} />;
  }

  const handleIconClick = () => {
    setPasswordVisible(!passwordVisible);
  };

  const { ref: removedRef, ...other } = props;

  // Set default format value for `tel` input
  if (type === 'tel' && !format) {
    format = '(###) ###-####';
  }

  // Set default placeholder value for `tel` input
  if (type === 'tel' && !placeholder) {
    placeholder = '(___) ___-____';
  }

  const content = (
    <div className={cnInputContainer}>
      {iconLeft}
      {format && (type === 'password' || type === 'tel' || type === 'text') && (
        <NumberFormat
          className={cnInput}
          format={format}
          mask="_"
          placeholder={placeholder}
          prefix={prefix}
          type={type}
          {...(props as NumberFormatProps)}
        />
      )}
      {type === 'currency' && (
        <NumberFormat
          allowLeadingZeros={false}
          className={cnInput}
          format={format}
          placeholder={placeholder ?? `${prefix ?? '$'}0.00`}
          prefix={prefix ?? '$'}
          thousandSeparator={true}
          type="text"
          {...(props as NumberFormatProps)}
        />
      )}
      {!format && type !== 'currency' && (
        <input
          {...(props as DetailedHTMLProps<unknown, unknown>)}
          type={type === 'password' && passwordVisible ? 'text' : type}
          placeholder={placeholder}
          className={cnInput}
          ref={ref}
        />
      )}
      {type === 'password' && (
        <Button
          classNames={[classInputIcon, `${classInputIcon}-hide-icon`]}
          iconColor={iconColorState}
          iconLeft={passwordVisible ? 'action_eye_closed' : 'action_eye_open'}
          onClick={handleIconClick}
          variant="tertiary"
        />
      )}
      {type !== 'password' && iconRight}
    </div>
  );

  return content;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(InputComponent);
