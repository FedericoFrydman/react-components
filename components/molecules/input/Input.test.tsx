import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Input, InputProps, InputType, InputValidationStatus } from './Input';
import { render, screen } from '@testing-library/react';
import { IconNames } from '../../../components/atoms/icon/Icon';

describe('Input Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Input name={testId} />);
    testId = 'test input';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the input the provided classNames', () => {
          const input = renderInput({ classNames });

          expect(input).toHaveClass(classNames.join(' '), { exact: false });
        });
      });
    });

    describe('disabled', () => {
      let disabled: boolean;

      it('defaults to enabled', () => {
        expect(renderInput()).toBeEnabled();
      });

      describe('when true', () => {
        beforeEach(() => (disabled = true));

        it('disables the input', () => {
          expect(renderInput({ disabled })).toBeDisabled();
        });
      });

      describe('when false', () => {
        beforeEach(() => (disabled = false));

        it('the input is enabled', () => {
          expect(renderInput({ disabled })).toBeEnabled();
        });
      });
    });

    describe('icon', () => {
      const iconClass = `lsux-input__icon`;
      const iconDefaultColor = 'BRAND200';
      const iconColorClass = `lsux-icon--${iconDefaultColor}`;
      let iconName: IconNames;

      describe('when "icon" is present', () => {
        beforeEach(() => (iconName = 'edit_add_plus'));

        it(`gives the icon the ${iconClass} class`, () => {
          renderInput({ icon: iconName });

          const icon = screen.getByRole('img');
          expect(icon).toHaveClass(iconClass);
        });

        it(`defaults the icon the color to ${iconColorClass} and gives the icon the class ${iconColorClass} class`, () => {
          renderInput({ icon: iconName });

          const icon = screen.getByRole('img');
          expect(icon).toHaveClass(iconColorClass);
        });

        it('passes the icon name to the icon', () => {
          renderInput({ icon: iconName });

          const icon = screen.getByRole('img');
          expect(icon).toHaveAttribute('alt', iconName + '.');
        });
      });

      describe('when "icon" is not present', () => {
        it('the input does not have the icon', () => {
          renderInput();

          const icon = screen.queryByRole('img');
          expect(icon).toBeNull();
        });
      });
    });

    describe('type', () => {
      describe('when provided a valid input type', () => {
        const inputTypes: InputType[] = [
          'color',
          'date',
          'datetime-local',
          'email',
          'hidden',
          'month',
          'number',
          'password',
          'range',
          'search',
          'tel',
          'text',
          'time',
          'url',
          'week',
          'currency',
        ];

        function itAssignsTheRightInputType(type: InputType) {
          describe(`when the type is ${type}`, () => {
            it(`sets the type html attribute as ${type}`, () => {
              const input = renderInput({ type });

              if (type === 'currency') type = 'text';

              expect(input).toHaveAttribute('type', type);
            });
          });
        }

        inputTypes.forEach((type) => itAssignsTheRightInputType(type));
      });

      describe('when not provided an input type', () => {
        it('defaults to the "text" type', () => {
          expect(renderInput()).toHaveAttribute('type', 'text');
        });
      });
    });

    describe('status', () => {
      const variants: InputValidationStatus[] = ['valid', 'warning', 'invalid'];

      it('defaults to no status', () => {
        const input = renderInput();

        expect(input).not.toHaveClass(`lsux-input-container__input--`);
      });

      function itAssignsTheClassForStatus(status: InputValidationStatus) {
        const inputStatusClass = `lsux-input-container__input--${status}`;

        describe(`when the status is ${status}`, () => {
          it(`assigns the ${inputStatusClass} to the input`, () => {
            const nput = renderInput({ status });

            expect(nput).toHaveClass(inputStatusClass);
          });
        });
      }

      variants.forEach((variant) => itAssignsTheClassForStatus(variant));
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLInputElement>();

    it('provides a ref to the input', () => {
      expect(renderInput({ ref })).toEqual(ref.current);
    });
  });

  describe('unspecified html attributes', () => {
    const altText = 'My accessible input';

    it('allows html attributes to be passed to the underlying input element', () => {
      render(<Input name={testId} alt={altText} data-testid={testId} />);

      const input = screen.getByTestId(testId);

      expect(input).toHaveAttribute('alt', altText);
    });
  });

  const renderInput = (props: InputProps = {}): HTMLElement => {
    render(<Input {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
