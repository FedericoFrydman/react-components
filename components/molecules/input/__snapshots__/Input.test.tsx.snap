// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should render the component 1`] = `
<div
  className="lsux-input-container"
>
  <input
    className="lsux-input-container__input"
    type="text"
  />
</div>
`;
